File: tidal-dl/server/routers/workers.py
"""Worker 注册 + 心跳 API"""
from fastapi import APIRouter, Depends
from pydantic import BaseModel
from typing import Optional
from database import get_db_dependency
from config import get_config

router = APIRouter(prefix="/api/workers", tags=["Workers"])


class WorkerRegister(BaseModel):
    name: str = ""
    hostname: str = ""
    ip: str = ""
    max_concurrency: int = 10


class WorkerHeartbeat(BaseModel):
    status: str = "online"
    active_tasks: int = 0
    total_downloaded: int = 0
    total_failed: int = 0
    total_bytes: int = 0


@router.get("")
def list_workers(db=Depends(get_db_dependency)):
    """获取所有 Worker"""
    cursor = db.cursor()
    cursor.execute("SELECT * FROM workers ORDER BY last_heartbeat DESC")
    return cursor.fetchall()


class WorkerUpdate(BaseModel):
    max_concurrency: Optional[int] = None
    name: Optional[str] = None


@router.put("/{worker_id}")
def update_worker(worker_id: str, data: WorkerUpdate, db=Depends(get_db_dependency)):
    """更新 Worker 配置"""
    cursor = db.cursor()
    updates = []
    params = []
    if data.max_concurrency is not None:
        updates.append("max_concurrency = %s")
        params.append(data.max_concurrency)
    if data.name is not None:
        updates.append("name = %s")
        params.append(data.name)
    if not updates:
        return {"message": "无修改"}
    params.append(worker_id)
    cursor.execute(f"UPDATE workers SET {', '.join(updates)} WHERE id = %s", params)
    db.commit()
    return {"message": "Worker 配置已更新"}


@router.delete("/{worker_id}")
def delete_worker(worker_id: str, db=Depends(get_db_dependency)):
    """删除 Worker"""
    cursor = db.cursor()
    cursor.execute("DELETE FROM workers WHERE id = %s", (worker_id,))
    db.commit()
    return {"message": "Worker 已删除"}


@router.post("/register")
def register_worker(data: WorkerRegister, db=Depends(get_db_dependency)):
    """Worker 注册（同 hostname+name 复用已有记录）"""
    import uuid
    cursor = db.cursor()

    # 查找已有同名 Worker
    cursor.execute(
        "SELECT id FROM workers WHERE hostname = %s AND name = %s LIMIT 1",
        (data.hostname, data.name or data.hostname)
    )
    existing = cursor.fetchone()

    if existing:
        worker_id = existing["id"]
        cursor.execute(
            "UPDATE workers SET ip = %s, status = 'online', "
            "last_heartbeat = NOW() WHERE id = %s",
            (data.ip, worker_id)
        )
    else:
        worker_id = f"w-{uuid.uuid4().hex[:12]}"
        cursor.execute(
            "INSERT INTO workers (id, name, hostname, ip, max_concurrency, status, last_heartbeat, registered_at) "
            "VALUES (%s, %s, %s, %s, %s, 'online', NOW(), NOW())",
            (worker_id, data.name or data.hostname, data.hostname, data.ip, data.max_concurrency)
        )

    # 分配一个可用账号（如果还没有）
    cursor.execute("SELECT assigned_account_id FROM workers WHERE id = %s", (worker_id,))
    w = cursor.fetchone()
    if not w or not w.get("assigned_account_id"):
        cursor.execute(
            "SELECT id FROM tidal_accounts WHERE status = 'active' ORDER BY total_downloads ASC LIMIT 1"
        )
        account = cursor.fetchone()
        if account:
            cursor.execute(
                "UPDATE workers SET assigned_account_id = %s WHERE id = %s",
                (account["id"], worker_id)
            )

    db.commit()

    # 返回完整配置
    config = get_config(db)
    cursor.execute("SELECT assigned_account_id FROM workers WHERE id = %s", (worker_id,))
    w = cursor.fetchone()
    return {
        "worker_id": worker_id,
        "config": config,
        "account_id": w["assigned_account_id"] if w else None
    }


@router.post("/{worker_id}/heartbeat")
def heartbeat(worker_id: str, data: WorkerHeartbeat, db=Depends(get_db_dependency)):
    """Worker 心跳"""
    cursor = db.cursor()
    cursor.execute(
        "UPDATE workers SET status = %s, active_tasks = %s, "
        "total_downloaded = %s, total_failed = %s, total_bytes = %s, "
        "last_heartbeat = NOW() WHERE id = %s",
        (data.status, data.active_tasks,
         data.total_downloaded, data.total_failed, data.total_bytes,
         worker_id)
    )
    db.commit()
    return {"message": "ok"}


@router.get("/{worker_id}/config")
def get_worker_config(worker_id: str, db=Depends(get_db_dependency)):
    """获取 Worker 配置"""
    cursor = db.cursor()
    cursor.execute("SELECT * FROM workers WHERE id = %s", (worker_id,))
    worker = cursor.fetchone()
    if not worker:
        return {"error": "Worker 不存在"}

    config = get_config(db)

    # 并发优先级: Worker 节点设置 > 全局配置
    worker_concurrency = worker.get("max_concurrency")
    effective_concurrency = worker_concurrency or config["download"]["concurrency"]

    return {
        "worker_id": worker_id,
        "concurrency": effective_concurrency,
        "quality": config["download"]["quality"],
        "proxy": config["proxy"],
        "s3": config["s3"],
        "download": config["download"],
    }

File: tidal-go/internal/server/handlers/workers.go
package handlers

import (
	"crypto/rand"
	"encoding/hex"
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"tidal-go/internal/models"
)

type registerReq struct {
	Name           string `json:"name"`
	Hostname       string `json:"hostname"`
	IP             string `json:"ip"`
	MaxConcurrency int    `json:"max_concurrency"`
}

// RegisterWorker 注册 worker(同 hostname+name 复用记录)。
func (h *Handler) RegisterWorker(c *gin.Context) {
	var req registerReq
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	name := req.Name
	if name == "" {
		name = req.Hostname
	}

	var existingID string
	err := h.DB.QueryRow("SELECT id FROM workers WHERE hostname=? AND name=? LIMIT 1", req.Hostname, name).Scan(&existingID)
	var workerID string
	if err == nil && existingID != "" {
		workerID = existingID
		_, _ = h.DB.Exec("UPDATE workers SET ip=?, status='online', last_heartbeat=NOW() WHERE id=?", req.IP, workerID)
	} else {
		workerID = "w-" + randHex(12)
		conc := req.MaxConcurrency
		if conc <= 0 {
			conc = 10
		}
		_, _ = h.DB.Exec(
			"INSERT INTO workers (id,name,hostname,ip,max_concurrency,status,last_heartbeat,registered_at) "+
				"VALUES (?,?,?,?,?,'online',NOW(),NOW())",
			workerID, name, req.Hostname, req.IP, conc,
		)
	}
	c.JSON(http.StatusOK, gin.H{"worker_id": workerID})
}

type heartbeatReq struct {
	Status          string `json:"status"`
	ActiveTasks     int    `json:"active_tasks"`
	TotalDownloaded int    `json:"total_downloaded"`
	TotalFailed     int    `json:"total_failed"`
	TotalBytes      int64  `json:"total_bytes"`
}

// Heartbeat worker 心跳。
func (h *Handler) Heartbeat(c *gin.Context) {
	workerID := c.Param("id")
	var req heartbeatReq
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	_, _ = h.DB.Exec(
		"UPDATE workers SET status=?, active_tasks=?, total_downloaded=?, total_failed=?, total_bytes=?, last_heartbeat=NOW() WHERE id=?",
		req.Status, req.ActiveTasks, req.TotalDownloaded, req.TotalFailed, req.TotalBytes, workerID,
	)
	c.JSON(http.StatusOK, gin.H{"message": "ok"})
}

// WorkerConfig 下发配置给 worker。
func (h *Handler) WorkerConfig(c *gin.Context) {
	workerID := c.Param("id")
	var w models.Worker
	if err := h.DB.Get(&w, "SELECT * FROM workers WHERE id=?", workerID); err != nil {
		c.JSON(http.StatusOK, gin.H{"error": "Worker 不存在"})
		return
	}
	cfg := h.Config()
	conc := w.MaxConcurrency
	if conc <= 0 {
		conc = cfg.Download.Concurrency
	}
	c.JSON(http.StatusOK, gin.H{
		"worker_id":   workerID,
		"concurrency": conc,
		"quality":     cfg.Download.Quality,
		"proxy":       cfg.Proxy,
		"s3":          cfg.EnabledS3(),
		"download":    cfg.Download,
	})
}

type updateWorkerReq struct {
	Name           *string `json:"name"`
	MaxConcurrency *int    `json:"max_concurrency"`
}

// UpdateWorker 管理端更新 worker(并发数等)。worker 每 30s 同步一次配置,热生效无需重启。
func (h *Handler) UpdateWorker(c *gin.Context) {
	workerID := c.Param("id")
	var req updateWorkerReq
	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": err.Error()})
		return
	}
	if req.MaxConcurrency != nil {
		if *req.MaxConcurrency < 1 || *req.MaxConcurrency > 10000 {
			c.JSON(http.StatusBadRequest, gin.H{"error": "max_concurrency 必须在 1-10000 之间"})
			return
		}
		if _, err := h.DB.Exec("UPDATE workers SET max_concurrency=? WHERE id=?", *req.MaxConcurrency, workerID); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
	}
	if req.Name != nil && *req.Name != "" {
		if _, err := h.DB.Exec("UPDATE workers SET name=? WHERE id=?", *req.Name, workerID); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
			return
		}
	}
	c.JSON(http.StatusOK, gin.H{"message": "ok"})
}

// ListWorkers 管理端列表。心跳超过 90s 视为 offline(status 字段是心跳时写入的,进程死掉不会自己更新)。
func (h *Handler) ListWorkers(c *gin.Context) {
	var workers []models.Worker
	_ = h.DB.Select(&workers, "SELECT * FROM workers ORDER BY last_heartbeat DESC")
	for i := range workers {
		if workers[i].LastHeartbeat == nil || time.Since(*workers[i].LastHeartbeat) > 90*time.Second {
			workers[i].Status = "offline"
		}
	}
	c.JSON(http.StatusOK, workers)
}

// DeleteWorker 删除 worker 记录(管理端)。
func (h *Handler) DeleteWorker(c *gin.Context) {
	_, _ = h.DB.Exec("DELETE FROM workers WHERE id=?", c.Param("id"))
	c.JSON(http.StatusOK, gin.H{"message": "Worker 已删除"})
}

func randHex(n int) string {
	b := make([]byte, n/2)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}
